src/Microsoft.Health.Fhir.Shared.Api/Features/Resources/ProfileResourcesBehaviour.cs

```diff
@@ -0,0 +1,63 @@
+using EnsureThat;
+using MediatR;
+using Microsoft.Health.Fhir.Core.Exceptions;
+using Microsoft.Health.Fhir.Core.Features.Security;
+using Microsoft.Health.Fhir.Core.Features.Security.Authorization;
+using Microsoft.Health.Fhir.Core.Messages.Create;
+using Microsoft.Health.Fhir.Core.Messages.Delete;
+using Microsoft.Health.Fhir.Core.Messages.Upsert;
+
+namespace Microsoft.Health.Fhir.Api.Features.Resources
+{
+    public sealed class ProfileResourcesBehaviour :
+        IPipelineBehavior<CreateResourceRequest, UpsertResourceResponse>,
+        IPipelineBehavior<UpsertResourceRequest, UpsertResourceResponse>,
+        IPipelineBehavior<ConditionalCreateResourceRequest, UpsertResourceResponse>,
+        IPipelineBehavior<ConditionalUpsertResourceRequest, UpsertResourceResponse>,
+        IPipelineBehavior<DeleteResourceRequest, DeleteResourceResponse>
+    {
+        private IFhirAuthorizationService _authorizationService;
+        private static IEnumerable<string> _supportedTypes = new HashSet<string>(StringComparer.OrdinalIgnoreCase) { "ValueSet", "StructureDefinition", "CodeSystem", "ConceptMap" };
+
+        public ProfileResourcesBehaviour(IFhirAuthorizationService authorizationService)
+        {
+            EnsureArg.IsNotNull(authorizationService, nameof(authorizationService));
+            _authorizationService = authorizationService;
+        }
+
+        public async Task<UpsertResourceResponse> Handle(ConditionalUpsertResourceRequest request, CancellationToken cancellationToken, RequestHandlerDelegate<UpsertResourceResponse> next)
+            => await GenericHandle(request.Resource.InstanceType, next);
+
+        public async Task<UpsertResourceResponse> Handle(ConditionalCreateResourceRequest request, CancellationToken cancellationToken, RequestHandlerDelegate<UpsertResourceResponse> next)
+            => await GenericHandle(request.Resource.InstanceType, next);
+
+        public async Task<UpsertResourceResponse> Handle(UpsertResourceRequest request, CancellationToken cancellationToken, RequestHandlerDelegate<UpsertResourceResponse> next)
+            => await GenericHandle(request.Resource.InstanceType, next);
+
+        public async Task<UpsertResourceResponse> Handle(CreateResourceRequest request, CancellationToken cancellationToken, RequestHandlerDelegate<UpsertResourceResponse> next)
+            => await GenericHandle(request.Resource.InstanceType, next);
+
+        public async Task<DeleteResourceResponse> Handle(DeleteResourceRequest request, CancellationToken cancellationToken, RequestHandlerDelegate<DeleteResourceResponse> next)
+            => await GenericHandle(request.ResourceKey.ResourceType, next);
+
+        private async Task<TResponse> GenericHandle<TResponse>(string resourceType, RequestHandlerDelegate<TResponse> next)
+        {
+            if (_supportedTypes.Contains(resourceType) && await _authorizationService.CheckAccess(DataActions.ProfileAdmin) != DataActions.ProfileAdmin)
```

Comment:
@CaitlinV39 we should definitely document this. Editing `"ValueSet", "StructureDefinition", "CodeSystem", "ConceptMap"` required this new permission.

Reply:
I need to remove "ConceptMap" from this list (I actually have that list as static collection in my other PR, just need to merge it and reuse), but, it's need to be documented, I agree.